File: js/client.test.js
const { test, describe } = require('node:test');
const assert = require('node:assert');
const http = require('node:http');
const APIClient = require('./client');

function createTestServer() {
  return http.createServer((req, res) => {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ status: 'ok' }));
  });
}

describe('APIClient', () => {
  test('should perform a successful request', async () => {
    const server = createTestServer();
    await new Promise(resolve => server.listen(0, resolve));
    const port = server.address().port;
    const client = new APIClient(`http://localhost:${port}`);

    const res = await client.do({ method: 'GET', path: '/test' });
    
    assert.strictEqual(res.statusCode, 200);
    assert.strictEqual(res.error, null);
    
    server.close();
  });

  test('should perform concurrent requests', async () => {
    const server = createTestServer();
    await new Promise(resolve => server.listen(0, resolve));
    const port = server.address().port;
    const client = new APIClient(`http://localhost:${port}`);

    const requests = [
      { method: 'GET', path: '/1' },
      { method: 'GET', path: '/2' }
    ];

    const results = await client.doConcurrent(requests);
    
    assert.strictEqual(results.length, 2);
    results.forEach(res => {
      assert.strictEqual(res.statusCode, 200);
    });
    
    server.close();
  });

  test('should handle timeouts', async () => {
    const server = http.createServer((req, res) => {
      setTimeout(() => {
        res.writeHead(200);
        res.end();
      }, 100);
    });
    await new Promise(resolve => server.listen(0, resolve));
    const port = server.address().port;
    
    // Short timeout of 20ms
    const client = new APIClient(`http://localhost:${port}`, 20);
    const res = await client.do({ method: 'GET', path: '/timeout' });
    
    assert.ok(res.error, 'Expected a timeout error');
    assert.strictEqual(res.statusCode, 0);
    
    server.close();
  });

  test('should send headers and body correctly', async () => {
    let receivedHeaders = {};
    let receivedBody = '';

    const server = http.createServer((req, res) => {
      receivedHeaders = req.headers;
      req.on('data', chunk => { receivedBody += chunk; });
      req.on('end', () => {
        res.writeHead(200);
        res.end();
      });
    });
    await new Promise(resolve => server.listen(0, resolve));
    const port = server.address().port;
    const client = new APIClient(`http://localhost:${port}`);

    await client.do({
      method: 'POST',
      path: '/post',
      headers: { 'x-test': 'value' },
      body: { foo: 'bar' }
    });

    assert.strictEqual(receivedHeaders['x-test'], 'value');
    assert.strictEqual(receivedHeaders['content-type'], 'application/json');
    assert.strictEqual(receivedBody, JSON.stringify({ foo: 'bar' }));

    server.close();
  });

  test('should handle empty requests list gracefully', async () => {
    const client = new APIClient('http://localhost:0');
    const results = await client.doConcurrent([]);
    assert.strictEqual(results.length, 0);
  });

  test('should respect throttleLimit concurrency constraint', async () => {
    let active = 0;
    let maxActive = 0;
    const server = http.createServer((req, res) => {
      active++;
      if (active > maxActive) {
        maxActive = active;
      }
      setTimeout(() => {
        active--;
        res.writeHead(200);
        res.end();
      }, 20);
    });
    await new Promise(resolve => server.listen(0, resolve));
    const port = server.address().port;
    const client = new APIClient(`http://localhost:${port}`);
    const requests = [
      { method: 'GET', path: '/1' },
      { method: 'GET', path: '/2' },
      { method: 'GET', path: '/3' },
      { method: 'GET', path: '/4' }
    ];
    await client.doConcurrent(requests, 2);
    assert.ok(maxActive <= 2);
    assert.ok(maxActive > 0);
    server.close();
  });
});

File: go/client_test.go
package main

import (
	"context"
	"net/http"
	"net/http/httptest"
	"sync"
	"testing"
	"time"
)

func TestAPIClient_Do(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
		w.Write([]byte(`{"status":"ok"}`))
	}))
	defer server.Close()

	client := NewAPIClient(server.URL, 5*time.Second)
	res := client.Do(context.Background(), RequestParams{
		Method: http.MethodGet,
		Path:   "/test",
	})

	if res.StatusCode != http.StatusOK {
		t.Errorf("Expected status 200, got %d", res.StatusCode)
	}
}

func TestAPIClient_DoConcurrent(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(10 * time.Millisecond)
		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	client := NewAPIClient(server.URL, 5*time.Second)
	requests := []RequestParams{
		{Method: http.MethodGet, Path: "/1"},
		{Method: http.MethodGet, Path: "/2"},
		{Method: http.MethodGet, Path: "/3"},
	}

	results := client.DoConcurrent(context.Background(), requests, 2)

	if len(results) != 3 {
		t.Errorf("Expected 3 results, got %d", len(results) )
	}

	for _, res := range results {
		if res.StatusCode != http.StatusOK {
			t.Errorf("Expected status 200, got %d", res.StatusCode)
		}
	}
}

func TestAPIClient_Timeout(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(100 * time.Millisecond)
		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	// Set a very short timeout
	client := NewAPIClient(server.URL, 20*time.Millisecond)
	res := client.Do(context.Background(), RequestParams{
		Method: http.MethodGet,
		Path:   "/timeout",
	})

	if res.Error == nil {
		t.Error("Expected timeout error, got nil")
	}
}

func TestAPIClient_HeadersAndBody(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("X-Test") != "value" {
			t.Errorf("Expected header X-Test: value, got %s", r.Header.Get("X-Test"))
		}
		if r.Header.Get("Content-Type") != "application/json" {
			t.Errorf("Expected Content-Type: application/json, got %s", r.Header.Get("Content-Type"))
		}
		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()

	client := NewAPIClient(server.URL, 5*time.Second)
	client.Do(context.Background(), RequestParams{
		Method:  http.MethodPost,
		Path:    "/post",
		Headers: map[string]string{"X-Test": "value"},
		Body:    map[string]string{"foo": "bar"},
	})
}

func TestAPIClient_DoConcurrent_Throttling(t *testing.T) {
	var mu sync.Mutex
	active, maxActive := 0, 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		active++
		if active > maxActive {
			maxActive = active
		}
		mu.Unlock()
		time.Sleep(20 * time.Millisecond)
		mu.Lock()
		active--
		mu.Unlock()
		w.WriteHeader(http.StatusOK)
	}))
	defer server.Close()
	client := NewAPIClient(server.URL, 5*time.Second)
	requests := []RequestParams{
		{Method: http.MethodGet, Path: "/1"},
		{Method: http.MethodGet, Path: "/2"},
		{Method: http.MethodGet, Path: "/3"},
		{Method: http.MethodGet, Path: "/4"},
	}
	client.DoConcurrent(context.Background(), requests, 2)
	if maxActive > 2 {
		t.Errorf("Expected max concurrent requests to be at most 2, got %d", maxActive)
	}
	if maxActive == 0 {
		t.Error("Expected some active requests, got 0")
	}
}
